### PR TITLE
Fix deadline option

### PR DESCRIPTION
**Pull request by inaod568**
Fixes the deadline option, there was apparently some crashes reported by the testers. 
While at it I did some refactoring so that it became a bit more maintainable. Please look trough it thoroughly and test as much as you can and have time for. The tests pass so it should be fine but also have a look at if the tests need to be improved on etc. 

- [x] Tests passing
- [x] Black formatting
- [ - ] Migrations for any changes to the database schema
- [x] Rebase/merge the `dev` branch
- [ - ] Note in the CHANGELOG

### CODE REVIEW

File: dds_web/api/project.py [hunk 1]
```diff
@@ -196,20 +149,172 @@ def post(self):
...
+        )
+
+        return models.ProjectStatuses(status="Deleted", date_created=current_time), delete_message
+
+    def archive_project(
+        self, project: models.Project, current_time: datetime.datetime, aborted: bool = False
+    ):
+        """Archive project: Make status Archived.
+
+        Only possible from Available and Expired.
```
Comment (alneberg) [inline dds_web/api/project.py:289]: From a quick glance, it looks like you're allowing it to be archived from "In Progress" as well. Not sure how it's supposed to be?

Comment (inaod568) [reply dds_web/api/project.py:289]: I can change this. You can only abort. Which is checked below in
```
if not aborted and project.current_status == "In Progress":
```

Comment (inaod568) [reply dds_web/api/project.py:289]: Hmm I'll check. It's suppose to only be possible to abort.

File: dds_web/api/project.py [hunk 2]
```diff
@@ -196,20 +149,172 @@ def post(self):
...
+        """
+        # Check if valid status transition
+        self.check_transition_possible(current_status=project.current_status, new_status="Archived")
+
+        if not aborted and project.current_status == "In Progress":
+            if not (project.has_been_available and aborted):
+                raise DDSArgumentError(
+                    f"You cannot archive a project that has been made available previously. "
+                    "Abort the project if you wish to proceed."
+                )
```
Comment (inaod568) [inline dds_web/api/project.py:299]:
```suggestion
        if project.current_status == "In Progress":
            if not (project.has_been_available and aborted):
                raise DDSArgumentError(
                    f"You cannot archive a project that has been made available previously. "
                    "Abort the project if you wish to proceed."
                )
```

File: dds_web/api/project.py (hunk 1 above)
Comment (inaod568) [reply dds_web/api/project.py:289]: I mean you can Archive, Archived has a `is_aborted` column.

Comment (inaod568) [reply dds_web/api/project.py:289]: So it's actually aborting that's allowed from In Progress

Comment (inaod568) [reply dds_web/api/project.py:289]: But yes, I think I need to change somehting in the code below

File: dds_web/api/project.py [hunk 3]
```diff
@@ -196,20 +149,172 @@ def post(self):
...
+        )
+
+        return models.ProjectStatuses(status="In Progress", date_created=current_time)
+
+    def expire_project(
+        self, project: models.Project, current_time: datetime.datetime, deadline_in: int
+    ) -> models.ProjectStatuses:
+        """Expire project: Make status Expired.
+
+        Only possible Available.
```
Comment (aanil) [inline dds_web/api/project.py:240]:
```suggestion
        Only possible from Available.
```

File: dds_web/api/project.py (hunk 1 above)
Comment (inaod568) [reply dds_web/api/project.py:288]: Changed.

File: dds_web/api/project.py [hunk 4]
```diff
@@ -196,20 +146,166 @@ def post(self):
...
+
+        # Delete metadata from project row
+        self.delete_project_info(proj=project)
+        delete_message = (
+            f"\nAll files in project '{project.public_id}' deleted and project info cleared"
+        )
+
+        return models.ProjectStatuses(status="Deleted", date_created=current_time), delete_message
+
+    def archive_project(
```
Comment (aanil) [inline dds_web/api/project.py:281]: We can archive a project directly from 'In Progress' without the `is_aborted` flag ever being set. Is that what we wish to happen?

Comment (inaod568) [reply dds_web/api/project.py:281]: Check again, I readded the check

File: tests/test_project_status.py [hunk 5]
```diff
@@ -161,23 +161,9 @@ def test_aborted_project(module_client, boto3_session):
...
-
-    assert response.status_code == http.HTTPStatus.BAD_REQUEST
     assert project.current_status == "In Progress"
-    assert (
-        "Project cannot be archived from this status but can be aborted if it has ever been made available"
-        in response.json["message"]
-    )
 
-    new_status["new_status"] = "Available"
+    new_status = {"new_status": "Available"}
```
Comment (inaod568) [inline tests/test_project_status.py:166]:
```suggestion
    new_status = {"new_status": "Archived"}
    response = module_client.post(
        tests.DDSEndpoint.PROJECT_STATUS,
        headers=tests.UserAuth(tests.USER_CREDENTIALS["unitadmin"]).token(module_client),
        query_string={"project": project_id},
        json=new_status,
    )

    assert response.status_code == http.HTTPStatus.BAD_REQUEST
    assert project.current_status == "In Progress"
    assert (
        "Project cannot be archived from this status but can be aborted if it has ever been made available"
        in response.json["message"]
    )

    new_status["new_status"] = "Available"
```

**Review by aanil — APPROVED** [review]
Looks ok to me!

**Pull request merged**